### PR TITLE
Make BytesReference an interface

### PR DESCRIPTION
**Pull request by Tim-Brooks**
BytesReference is currently an abstract class which is extended by
various implementations. This makes it very difficult to use the
delegation pattern. The implication of this is that our releasable
BytesReference is a PagedBytesReference type and cannot be used as a
generic releasable bytes reference that delegates to any reference type.
This commit makes BytesReference an interface and introduces an
AbstractBytesReference for common functionality.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        String composite = "composite";
+        String paged = "paged";
+        String array = "array";
+        String type = randomFrom(composite, paged, array);
+        if (array.equals(type)) {
+            final BytesStreamOutput out = new BytesStreamOutput(length);
+            for (int i = 0; i < length; i++) {
+                out.writeByte((byte) random().nextInt(1 << 8));
+            }
+            assertEquals(length, out.size());
```
Comment (pugnascotia) [inline server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java]:
```suggestion
            assertThat(length, equalTo(out.size()));
```

File: server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            BytesArray ref = new BytesArray(out.bytes().toBytesRef().bytes, 0, length);
+            assertEquals(length, ref.length());
```
Comment (pugnascotia) [inline server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java]:
```suggestion
            assertThat(length, equalTo(ref.length()));
```

File: server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            assert composite.equals(type);
+            List<BytesReference> referenceList = new ArrayList<>();
+            for (int i = 0; i < length; ) {
+                int remaining = length - i;
+                int sliceLength = randomIntBetween(1, remaining);
+                ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(sliceLength, bigarrays);
+                for (int j = 0; j < sliceLength; j++) {
+                    out.writeByte((byte) random().nextInt(1 << 8));
+                }
+                assertEquals(sliceLength, out.size());
```
Comment (pugnascotia) [inline server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java]:
```suggestion
                assertThat(sliceLength, equalTo(out.size()));
```

File: server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                referenceList.add(out.bytes());
+                i += sliceLength;
+            }
+            BytesReference ref = new CompositeBytesReference(referenceList.toArray(new BytesReference[0]));
+            assertEquals(length, ref.length());
```
Comment (pugnascotia) [inline server/src/test/java/org/elasticsearch/common/bytes/ReleasableBytesReferenceTests.java]:
```suggestion
            assertThat(length, equalTo(ref.length()));
```

**Review by pugnascotia — APPROVED** [review]

**Review by pugnascotia — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/transport/CompressibleBytesOutputStream.java [hunk 5]
```diff
@@ -19,12 +19,12 @@
 
 package org.elasticsearch.transport;
 
-import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.compress.CompressorFactory;
 import org.elasticsearch.common.io.Streams;
 import org.elasticsearch.common.io.stream.BytesStream;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.internal.io.IOUtils;
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/transport/CompressibleBytesOutputStream.java]: NIT: revert this

**Review by original-brownbear — APPROVED** [review]
LGTM I'm all for this refactoring, thanks @tbrooks8

**Pull request merged**